### PR TITLE
planner: introduce a flag to cost calculation to control some behaviors

### PR DESCRIPTION
**Pull request by qw4990**
### What problem does this PR solve?

Issue Number: ref #33945

Problem Summary: planner: introduce a flag to cost calculation to control some behaviors

### What is changed and how it works?

No logical change, just refactor that introduce a flag to cost calculation to control some behaviors

### Check List

Tests 

- [x] Unit test
- [ ] Integration test
- [ ] Manual test (add detailed scripts or steps below)
- [ ] No code

Side effects

- [ ] Performance regression: Consumes more CPU
- [ ] Performance regression: Consumes more Memory
- [ ] Breaking backward compatibility

Documentation

- [ ] Affects user behaviors
- [ ] Contains syntax changes
- [ ] Contains variable changes
- [ ] Contains experimental features
- [ ] Changes MySQL compatibility

### Release note

Please refer to [Release Notes Language Style Guide](https://pingcap.github.io/tidb-dev-guide/contribute-to-tidb/release-notes-style-guide.html) to write a quality release note.

```release-note
None
```

### CODE REVIEW

File: planner/core/plan_cost.go [hunk 1]
```diff
@@ -26,15 +26,27 @@ import (
 	"github.com/pingcap/tidb/util/paging"
 )
 
+const (
+	// CostFlagRecalculate indicates the optimizer to ignore cached cost and recalculate it again.
+	CostFlagRecalculate uint64 = 1 << iota
+
+	// CostFlagUseTrueCardinality indicates the optimizer to use true cardinality to calculate the cost.
+	CostFlagUseTrueCardinality
```
Comment (qw4990) [inline planner/core/plan_cost.go:34]: In the next PR, I'll use these two flags to support `explain analyze format='true_card_cost' ...`.

File: planner/core/plan_cost.go [hunk 2]
```diff
@@ -1026,22 +1049,29 @@ func (p *PhysicalUnionAll) GetPlanCost(taskType property.TaskType) (float64, err
...
 	p.planCost = childCost
 	// accumulate net cost
 	// TODO: this formula is wrong since it doesn't consider tableRowSize, fix it later
-	p.planCost += p.children[0].StatsCount() * p.ctx.GetSessionVars().GetNetworkFactor(nil)
+	p.planCost += getStatsCount(p.children[0], costFlag) * p.ctx.GetSessionVars().GetNetworkFactor(nil)
 	p.planCostInit = true
 	return p.planCost, nil
 }
 
+func getStatsCount(operator PhysicalPlan, costFlag uint64) float64 {
```
Comment (time-and-fate) [inline planner/core/plan_cost.go:1068]: This function name is a bit ambiguous.

**Review by time-and-fate — APPROVED** [review]

File: planner/core/plan_cost.go (hunk 2 above)
Comment (qw4990) [reply planner/core/plan_cost.go]: OK, I renamed it to `getCardinality`

**Review by Reminiscent — APPROVED** [review]

**Pull request merged**